Bluetooth: host: Add support for retrieving the remote version

Blackladder opened the pull request:
Make remote features and remote version accesible to the application
through the bt_conn_get_info object.
get_remote_version() is called if the remote version is not stored
in the new fields in struct bt_conn.

Vudentz commented on include/bluetooth/conn.h:
```diff
@@ -162,6 +171,8 @@ struct bt_conn_info {
 
 		struct bt_conn_br_info br;
+
+		struct bt_conn_rv_info rv;
```
This is overlapping with le_info since it is a union, we need to store this info elsewhere, perhaps a dedicated API that does take a pointer to store this info and block while doing the HCI command, though in case of multiple thread trying to access it may have to be a lock/flag.

Blackladder replied:
@Vudentz Could I just move the remote version struct outside of the union and wrap it in a "#if defined"?

joerchan commented on include/bluetooth/conn.h:
```diff
@@ -129,13 +129,21 @@ struct bt_conn_le_info {
 	u16_t interval; /** Connection interval */
 	u16_t latency; /** Connection slave latency */
 	u16_t timeout; /** Connection supervision timeout */
+	u8_t *features; /** Remote device's ble features */
```
Declare this one as const u8_t * features, this will allow it to be assigned to a const object. Which is the case in:
int bt_conn_get_info(**const** struct bt_conn *conn, ...)

joerchan replied:
Also we should give some information about the length. I think we should create a `#define BT_LE_FEATURE_LEN 8` and reference that here.

joerchan commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1812,6 +1812,10 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
+#endif /* CONFIG_BT_TESTING */
 		return 0;
 #if defined(CONFIG_BT_BREDR)
 	case BT_CONN_TYPE_BR:
```
explicitly set info->rv = NULL here?
We don't initiate read_remote_version for BR/EDR connections.

Blackladder commented:
Added the changes mentioned by @joerchan

jhedberg commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -135,7 +135,10 @@ struct bt_conn {
-#endif
+#endif /* CONFIG_BT_BREDR */
+#if defined(CONFIG_BT_TESTING)
+		struct bt_conn_rv_info rv;
```
The union is for things which are mutually exclusive. Specifically, the idea is that the member used matches `conn->type`. It is therefore the wrong place to put bt_conn_rv_info.

Blackladder replied:
I actually fixed this like two weeks ago after Vudentz mentioned it, but I forgot to commit the change. It has now been pushed.

Vudentz commented:
If I recall correctly, I suggested having a dedicated API to trigger the remote info, i.e bt_conn_remote_info so we don't cause extra traffic with the likes of bt_conn_info.

Blackladder commented:
With the latest commit, I have separated the retrieval of remote version into a new function bt_conn_remote_info like @Vudentz suggested and there no longer is a rv struct in bt_conn_info.

Vudentz commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1823,6 +1824,20 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
+#if defined(CONFIG_BT_TESTING)
+int bt_conn_remote_info(const struct bt_conn *conn, struct bt_conn_rv_info *rv)
+{
+	if (conn->type == BT_CONN_TYPE_LE) {
```
Sorry but this is not quite right, the idea was that hci_read_remote_version would be called, if it hasn't been called yet, so that only application using this function would incur in extra latency fetching this info. Also that means you don't need to hide this functionality behing CONFIG_BT_TESTING.

Blackladder replied:
Fair point, but that is above my skill level.

Vudentz replied:
Not sure I understand you comment about skill level, what I suggested here is to check if rv has been read and if not call hci_read_remote_version which should use bt_hci_cmd_send_sync and block until the info is returned.

Blackladder replied:
What I mean is that I cannot make something that both does the task it is supposed to do and still follow the code standards.

I could for example use something like:
if (conn->rv == null) {
    hci_read_remote_version(conn);
    while (conn->rv == null) { /* or something with a timeout */
        k_sleep(/* some reasonable time */);
    }
} else {
    /* assign the values */
}

But I have a strong suspicion that this is not up to the standards set.

Vudentz replied:
Just make hci_read_remote_version use bt_hci_cmd_send_sync that one will block waiting the response on semaphore so you don't need to do any busy loop or anything:

https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/hci_core.c#L307

So all you need to do is replace bt_hci_cmd_send with bt_hci_cmd_send_sync, parse the response setting rv.

Blackladder replied:
I pushed the latest changes. As you mentioned the CONFIG_BT_TESTING would not be needed, I wondered if it should be wrapped in a broader config like CONFIG_BT_CONN or something.

Vudentz replied:
Yes, if it depends on things like bt_conn it should be conditional to CONFIG_BT_CONN if that is not already the case, but it looks like this is still not calling hci_read_remote_version directly so the change to bt_hci_cmd_send_sync is useless as the processing is asynchronous which is racy as the event may come after bt_conn_get_info is called thus the suggestion to use bt_hci_cmd_send_sync and process the event synchronously.

Blackladder replied:
Let me see if I get this right. bt_hci_cmd_send_sync only blocks until the command is executed, but does not wait for the event?

Vudentz replied:
bt_hci_cmd_send_sync only works with commands that respond via command complete, if the response comes as a separate event then we either have to have it to work with an arbritrary opcode or have a separate semaphore to block on until we wait the event to arrive.

Blackladder replied:
Do you have a suggestion how to share a semaphore between hci_read_remote_version and bt_hci_evt_read_remote_version_complete?

Blackladder replied:
Implemented with callback instead. Tested with my application and it worked.

wopu-ot commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2936,6 +2958,30 @@ static void hci_encrypt_key_refresh_complete(struct net_buf *buf)
+
+	if (!evt->status) {
+		conn->rv.version = evt->version;
+		conn->rv.manufacturer = evt->manufacturer;
```
I am pretty sure these fields need to be endianness-adjusted (cf. the conversion of evt->handle)

Blackladder replied:
I don't think so seeing as the example data I have received is only valid un-adjusted. My personal phone's bluetooth chip manufacturer 0x000f is a valid company identifier (Broadcom Corporation), while 0x0f00 is not valid. Highest company identifier is as of today "2012   0x07DC 'ThingOS GmbH'". https://www.bluetooth.com/specifications/assigned-numbers/company-identifiers/

wopu-ot replied:
I assume you run on a little-endian architecture? sys_le16_to_cpu() is a no-op on a little-endian architecture, but necessary on a big-endian architecture to read the correct 0x000f instead of 0x0f00.

Blackladder replied:
I think it's running in little-endian architecture, yes.
So you want this:?
´if (!evt->status) {
		conn->rv.version = sys_le16_to_cpu(evt->version);
		conn->rv.manufacturer = sys_le16_to_cpu(evt->manufacturer);
		conn->rv.subversion = sys_le16_to_cpu(evt->subversion);
	} ´

wopu-ot replied:
Yes, I believe that would be the correct version.

Blackladder replied:
Pushed the change now

Blackladder commented:
@Vudentz Can I get some help fixing the error found by shippable. It says that it runs out of memory compiling the test\bluetooth\mesh_shell for nrf51_pca10028. (it tries to use 100.01% of the memory) I tried compiling the same example for nrf51_pca10028 myself and it "only" used 98.05%.

Vudentz commented:
> @Vudentz Can I get some help fixing the error found by shippable. It says that it runs out of memory compiling the test\bluetooth\mesh_shell for nrf51_pca10028. (it tries to use 100.01% of the memory) I tried compiling the same example for nrf51_pca10028 myself and it "only" used 98.05%.

Try running sanity-check locally, i.e scripts/sanitycheck -p nrf51_pca10028 -t bluetooth, if that is reproducible there then we either have to remove nrf51from the list of platforms or tune the stack sizes, etc to make it fit again.

Vudentz commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -749,6 +753,28 @@ static int hci_le_read_remote_features(struct bt_conn *conn)
+	int err = bt_hci_cmd_send_sync(BT_HCI_OP_READ_REMOTE_VERSION_INFO, buf,
+				NULL);
+	if  (err != 0) {
+		return err;
```
If send fails you would left remote_version_cb set here so Id leave it to be set later, though there is another problem where another thread may call hci_read_remote_version in the meantime and override the cb so we may need a semaphore either way to ensure only one thread is pending waiting the callback at the time.

Blackladder replied:
So you want this?

```
if  (err != 0) {
    remote_version_cb = NULL;
    return err;
}
```

Blackladder replied:
@Vudentz Considering what the function does, how long it takes (or rather doesn't) and how few people are actually going to use it, is that really necessary?

Blackladder replied:
@Vudentz What would be a suitable timeout for k_sem_take waiting for get_remote version to clear? 1-5 seconds maybe? (I'm thinking K_FOREVER is a probably a bad idea)

Vudentz replied:
I guess z_ is for use on userspace code, though I may be wrong here since I never really use them. Back the the timeout, we could use the same interval as bt_hci_cmd_send_sync uses.

Blackladder commented:
Sanitycheck doesn't work on windows (which I'm using), but I tried to compile using the latest Zephyr version and it compiled using 99.99% of SRAM. So... there's the problem.

Blackladder commented:
@jhedberg Could you remove you change request seeing as we no longer store remote version in conn  anymore, we switched to callback. Feel free to request any other changes as you see fit. :-)

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -749,6 +754,34 @@ static int hci_le_read_remote_features(struct bt_conn *conn)
+int hci_read_remote_version(struct bt_conn *conn, bt_remote_version_cb *rv_cb)
+{
+	int err = k_sem_take(&remote_version_sem, HCI_CMD_TIMEOUT);
+	if (err != 0) {
```
Normal pattern would be like this:

```
if (err) {
    return err;
}

/* rest of the code */
```

joerchan commented on include/bluetooth/conn.h:
```diff
@@ -129,13 +129,26 @@ struct bt_conn_le_info {
+/** Remote Version callback
+ * @param info Structure containing the remote info of the peer
+ */
+typedef void (bt_remote_version_cb)(struct bt_conn_rv_info);
```
Would it be better to give a pointer here? (struct bt_conn_rv_info *)

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2936,6 +2969,25 @@ static void hci_encrypt_key_refresh_complete(struct net_buf *buf)
+		rv.manufacturer = sys_le16_to_cpu(evt->manufacturer);
+		rv.subversion = sys_le16_to_cpu(evt->subversion);
+	}
+	remote_version_cb(rv);
```
If evt->status is non-zero here how can the user know? I think it would be best to give a pointer to remote version, then we can give NULL if the procedure failed.

Blackladder replied:
I have now changed to pointers (not yet pushed), should I add a status field to the bt_conn_rv_info so that if there is an error the developer can know the error code?
Or should I just stick with giving NULL?

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -749,6 +754,34 @@ static int hci_le_read_remote_features(struct bt_conn *conn)
+
+		cp = net_buf_add(buf, sizeof(*cp));
+		cp->handle = sys_cpu_to_le16(conn->handle);
+		remote_version_cb = rv_cb;
```
Move this assignment under the err handling of checking the command?  Then you don't need to unassign it in the if-block.

Blackladder commented:
Implemented the changes requested by @joerchan .

carlescufi commented:
@Vudentz @jhedberg can you please re-review?

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3528,6 +3584,9 @@ static const struct event_handler normal_events[] = {
 		      sizeof(struct bt_hci_evt_disconn_complete)),
+	EVENT_HANDLER(BT_HCI_EVT_REMOTE_VERSION_INFO,
+			  bt_hci_evt_read_remote_version_complete,
+			  sizeof(struct bt_hci_evt_remote_version_info)),
```
The split lines don't seem to be properly aligned with the opening parenthesis of the first line

Blackladder replied:
This is a really strange error, because in my IDE (vscode) all of the event handlers are not aligned properly and my addition is aligned with the others.
Apparently vscode and github has different lengths of tabs?
The alignment on my addition was three tabs and two spaces while the alignment of the others are two tabs and six spaces.
For now I just fixed mine to match the others although there is no visual difference on my side (in my IDE).

Blackladder commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1823,6 +1824,17 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
+	}
+
+	return 0;
+}
```
The function used to have more in it, I'll remove the funtion and rename `hci_read_remote_version` as you mentioned and move the header reference from hci_core.h to hci.h.

Blackladder commented:
Fixed errors/inconsistencies mentioned by @jhedberg.

Blackladder commented on include/bluetooth/conn.h:
```diff
@@ -129,13 +129,26 @@ struct bt_conn_le_info {
+/** Remote Version callback
+ * @param info Structure containing the remote info of the peer
+ */
+typedef void (bt_remote_version_cb)(struct bt_conn_rv_info *);
```
Fix: name no longer missing

Blackladder commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -135,7 +135,7 @@ struct bt_conn {
 		struct bt_conn_br	br;
 		struct bt_conn_sco	sco;
-#endif
+#endif /* CONFIG_BT_BREDR */
```
Made changes in this file at an earlier point and I thought why not add the comment, seems kind of silly now just there on it's own.
Fix: removed inserted comment

Blackladder commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -749,6 +754,36 @@ static int hci_le_read_remote_features(struct bt_conn *conn)
 
+int hci_read_remote_version(struct bt_conn *conn, bt_remote_version_cb *rv_cb)
+{
+	int err = k_sem_take(&remote_version_sem, HCI_CMD_TIMEOUT);
```
Fix: Did as shown.

Blackladder commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -749,6 +754,36 @@ static int hci_le_read_remote_features(struct bt_conn *conn)
+		cp = net_buf_add(buf, sizeof(*cp));
+		cp->handle = sys_cpu_to_le16(conn->handle);
+		err = bt_hci_cmd_send_sync(BT_HCI_OP_READ_REMOTE_VERSION_INFO, buf,
+					NULL);
```
Aligned to `(` .

Blackladder commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -749,6 +754,36 @@ static int hci_le_read_remote_features(struct bt_conn *conn)
+			return err;
+		}
+
+		remote_version_cb = rv_cb;
```
Fix: Moved setting of callback to before `hci_cmd_send_sync`

Blackladder commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2936,6 +2971,27 @@ static void hci_encrypt_key_refresh_complete(struct net_buf *buf)
+		remote_version_cb(&rv);
+	} else {
+		remote_version_cb(NULL);
+	}
```
Fix: Inserted empty line.

jhedberg commented on include/bluetooth/hci.h:
```diff
@@ -1938,6 +1938,15 @@ typedef bool bt_hci_vnd_evt_cb_t(struct net_buf_simple *buf);
+ *
+ *  @return Zero on success or (negative) error code on failure.
+ */
+int bt_conn_remote_version(struct bt_conn *conn, bt_remote_version_cb *rv_cb);
```
Sorry, I should have said this earlier when you hinted at moving this here: the right place is `conn.h` as the name of the function also suggests (no "hci" in it). The Zephyr Bluetooth API is designed to be independent from HCI (with the exception of `hci.h`) in order to allow replacing the stack with a potentially non-HCI based one (e.g. the stack running on a separate chip with some high-level interface to it).

Blackladder replied:
So I undo the removal of the function and just shorten it in the manner you described earlier?
Or do you have another suggestion?

jhedberg replied:
Just move the declaration from hci.h to conn.h. No other change needed.

Blackladder replied:
The function is in hci_core.c, don't we need something else in order to have it in conn.h?

jhedberg replied:
Oh, then conn.c us more appropriate, but the header location is the most important part.

Blackladder replied:
The functions need to be in the same file since they share a semaphore, so I should move both functions to conn.c? Or should i just leave it as is and just have the header reference in conn.h with the functions in hci_core.c?

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -93,6 +93,11 @@ static struct bt_pub_key_cb *pub_key_cb;
 
+#if defined(CONFIG_BT_CONN)
+static bt_remote_version_cb *remote_version_cb;
+K_SEM_DEFINE(remote_version_sem, 1, 1);
```
Why do you need a semaphore?

Blackladder replied:
@Vudentz wanted it in case it was called several times in a short timeframe

jhedberg replied:
Hmm, I guess it'll deadlock then if it's called a second time from the RX thread (say, a connected callback)? Sorry that this PR has dragged on for so long, but I'd like to discuss some other options, such as making the event a priority one (i.e. having it processed through the priority RX thread). Maybe we could have a short chat with you and @Vudentz on Slack today?

Blackladder replied:
Sure, I'm available from about 13:00 (GMT +2), just tell me when.

jhedberg replied:
> Sure, I'm available from about 13:00 (GMT +2), just tell me when.

I've pinged you on #bluetooth earlier today. Are you available now @Blackladder ?

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1866,6 +1873,9 @@ const bt_addr_le_t *bt_conn_get_dst(const struct bt_conn *conn)
 
 int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
 {
+	if (conn->state == BT_CONN_CONNECTED) {
```
!=

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1866,6 +1873,9 @@ const bt_addr_le_t *bt_conn_get_dst(const struct bt_conn *conn)
 {
+	if (conn->state == BT_CONN_CONNECTED) {
+		return -ENOTCONN;
+	}
```
Add an empty line after the }

jhedberg commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -138,6 +141,14 @@ struct bt_conn {
 	};
+#if defined(CONFIG_BT_REMOTE_VERSION)
+	struct bt_conn_rv {
+	u8_t  version;
```
Wrong indentation here for the struct members (should be one more tab)

Blackladder replied:
Like this?
```
#if defined(CONFIG_BT_REMOTE_VERSION)
	struct bt_conn_rv {
		u8_t	version;
		u16_t	manufacturer;
		u16_t	subversion;
	struct k_sem rv_sem;
	} rv;
#endif
```

jhedberg replied:
No. Everything between `{ ... }` should have the same indentation. I.e. it should look something like this:
```
struct bt_conn_rv {
	u8_t  version;
	u16_t manufacturer;
	u16_t subversion;
	struct k_sem sem;
} rv;
```

Note that there's no reason to have a `rv_` prefix for the semaphore since this is already inside a struct called `rv`

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -914,6 +914,37 @@ static int hci_le_read_remote_features(struct bt_conn *conn)
+	cp = net_buf_add(buf, sizeof(*cp));
+	cp->handle = sys_cpu_to_le16(conn->handle);
+	atomic_set_bit(conn->flags, BT_CONN_PENDING_VERSION_INFO);
+	err = bt_hci_cmd_send(BT_HCI_OP_READ_REMOTE_VERSION_INFO, buf);
```
Use send_sync so you catch a potential non-zero command status

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -914,6 +914,37 @@ static int hci_le_read_remote_features(struct bt_conn *conn)
+	if (err) {
+		k_sem_give(&conn->rv.rv_sem);
+		return err;
+	}
```
Empty line after }

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3197,6 +3228,33 @@ static void hci_encrypt_key_refresh_complete(struct net_buf *buf)
+		conn->rv.version = evt->version;
+		conn->rv.manufacturer = sys_le16_to_cpu(evt->manufacturer);
+		conn->rv.subversion = sys_le16_to_cpu(evt->subversion);
+		atomic_clear_bit(conn->flags, BT_CONN_PENDING_VERSION_INFO);
```
Move this outside (before) the if-statement, since it's needed in both branches

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -914,6 +914,37 @@ static int hci_le_read_remote_features(struct bt_conn *conn)
+	struct bt_hci_cp_read_remote_version_info *cp;
+	struct net_buf *buf;
+
+	conn = bt_conn_lookup_handle(handle);
```
Don't do this - you already have the conn object in conn.c so a bit silly going from bt_conn to handle and then back to bt_conn. I'd just put the helper in conn.c. Not everything HCI related has to be in hci_conn.c. Just the main/core things (hence the name of the file).

Blackladder replied:
But the conn in `bt_conn_get_info` is const, and I got error that the const identifier was being discarded.

jhedberg replied:
What error exactly?

Blackladder replied:
I updated such that I have newest version of zephyr and now I get another error:
```
In file included from C:/ncs/zephyr/include/arch/arm/cortex_m/cmsis.h:17:0,
                 from C:/ncs/zephyr/include/arch/arm/cortex_m/mpu/arm_mpu_v7m.h:10,
                 from C:/ncs/zephyr/include/arch/arm/cortex_m/mpu/arm_mpu.h:13,
                 from C:/ncs/zephyr/include/arch/arm/arch.h:246,
                 from C:/ncs/zephyr/include/arch/cpu.h:17,
                 from C:/ncs/zephyr/include/kernel_includes.h:34,
                 from C:/ncs/zephyr/include/kernel.h:17,
                 from C:/ncs/zephyr/kernel/include/kernel_structs.h:10,
                 from C:/ncs/zephyr/arch/arm/core/offsets/offsets.c:26:
c:/ncs/zephyr/soc/arm/nordic_nrf/nrf52/soc.h:17:10: fatal error: nrf.h: No such file or directory
 #include <nrf.h>
```

joerchan replied:
@Blackladder You need to do 'west update'

jhedberg commented:
Any update on this one?

Blackladder committed:
Bluetooth: host: Add calling of read_remote_version

Make remote features and remote version accesible to the application
through the bt_conn_get_remote_info object. The host will auto initiate
the procedures. If the procedures have not finished with the application
calls bt_conn_get_remote_info then EBUSY will be returned.
The procedures should finish during the first 10 connection intervals.

Signed-off-by: Sverre Storvold <Sverre.Storvold@nordicsemi.no>
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Blackladder commented:
Regarding the area:samples thing: I added CONFIG_BT_REMOTE_VERSION to the bluetooth/peripheral sample and forgot to remove it before I pushed, but it's fixed now.

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -249,6 +249,14 @@ config BT_AUTO_PHY_UPDATE
 
+config BT_REMOTE_VERSION
+	bool "Enable fetching of remote version"
+	default n
```
The implicit default is `n` so you can remove this line

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -914,6 +914,36 @@ static int hci_le_read_remote_features(struct bt_conn *conn)
 }
 
+#if defined(CONFIG_BT_REMOTE_VERSION)
+int hci_le_read_remote_version(struct bt_conn *conn)
```
Move this into `conn.c`. There's no reason to have it in hci_core.c and then you have to export it through a header file instead of it being static.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3197,6 +3227,33 @@ static void hci_encrypt_key_refresh_complete(struct net_buf *buf)
+		conn->rv.version = evt->version;
+		conn->rv.manufacturer = sys_le16_to_cpu(evt->manufacturer);
+		conn->rv.subversion = sys_le16_to_cpu(evt->subversion);
+		atomic_clear_bit(conn->flags, BT_CONN_PENDING_VERSION_INFO);
```
You're duplicating this clearing in both branches. Just move it outside both of them, i.e. before the if-statment.

jhedberg commented on include/bluetooth/conn.h:
```diff
@@ -133,6 +133,8 @@ enum {
 };
 
+
+
```
These two added lines don't seem to be in any way related? Please remove them.

jhedberg commented on include/bluetooth/conn.h:
```diff
@@ -169,6 +172,7 @@ enum {
  *  @param id Which local identity the connection was created with
  *  @param le LE Connection specific Info
  *  @param br BR/EDR Connection specific Info
+ *  @param rv LE version info
```
This seems out of date. You're not adding any member called `rv` to the `bt_conn_info` struct.

jhedberg reviewed (changes requested):
Starting to look better, thanks. Still some comments though.

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1895,6 +1922,36 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
+	atomic_set_bit(conn->flags, BT_CONN_PENDING_VERSION_INFO);
+	err = bt_hci_cmd_send(BT_HCI_OP_READ_REMOTE_VERSION_INFO, buf);
+	if (err) {
+		k_sem_give(&conn->rv.sem);
```
Missing `atomic_clear_bit(conn->flags, BT_CONN_PENDING_VERSION_INFO);` in this branch

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1895,6 +1922,36 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
+	cp = net_buf_add(buf, sizeof(*cp));
+	cp->handle = sys_cpu_to_le16(conn->handle);
+	atomic_set_bit(conn->flags, BT_CONN_PENDING_VERSION_INFO);
+	err = bt_hci_cmd_send(BT_HCI_OP_READ_REMOTE_VERSION_INFO, buf);
```
Use `bt_hci_cmd_send_sync()` here - that way you wait for the Command Status HCI event and catch a potential error there (you'll get a non-zero return in that case). With the current PR if the HCI command would fail with a Command Status you'd still end up waiting the full HCI_CMD_TIMEOUT, which is not correct.

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1895,6 +1922,36 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
+	if (err) {
+		k_sem_give(&conn->rv.sem);
+		return err;
+	}
```
Empty line after `}`

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1884,6 +1893,24 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
+
+			if (err) {
+				return err;
+			} else {
```
The else-branch is unnecessary since you return from the first branch

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1895,6 +1922,36 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
 }
 
+#if defined(CONFIG_BT_REMOTE_VERSION)
+int hci_le_read_remote_version(struct bt_conn *conn)
```
Move this to `conn.c` as a static function. That way you don't need to touch `hci_core.h`.

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1895,6 +1922,36 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
+	struct bt_hci_cp_read_remote_version_info *cp;
+	struct net_buf *buf;
+
+	k_sem_init(&conn->rv.sem, 0, 1);
```
This doesn't look right. What if there's another `bt_conn_get_info()` call on the same connection that's blocking on `k_sem_take()`? You'd end up messing up the state of the semaphore by calling `k_sem_init()`. The initialisation of the semaphore should happen when the connection object is created, i.e. probably in `conn_new()`.

You're also missing protection for the multiple access like I described above. What might make most sense is to move the `atomic_set_bit(conn->flags, BT_CONN_PENDING_VERSION_INFO);` from below up here, and make it:

```
if (atomic_test_and_set_bit(conn->flags, BT_CONN_PENDING_VERSION_INFO)) {
        return -EBUSY;
}
```

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1643,6 +1646,12 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 			if (conn->type == BT_CONN_TYPE_LE) {
 				k_delayed_work_cancel(&conn->update_work);
+#if defined(CONFIG_BT_REMOTE_VERSION)
+	if (atomic_test_and_clear_bit(conn->flags,
```
The indentation isn't right here - it should be the same as all other code in the same block, regardless of the #ifdef being there or not.

Blackladder replied:
Yeah, one problem. If the "if" is to match the indentation and `BT_CONN_PENDING_VERSION_INFO` is to match the `atomic_test_and_clear_bit()` the line is over 80 characters.

jhedberg replied:
80 is not the hard rule - it'll cause a warning but shouldn't prevent merging. It does give a good hint to look at overall code design though, and in this case it seems the sane thing to do would be to refactor the `bt_conn_set_state()` function into smaller helper functions for each state. Anyway, that's something for a separate PR. Main point is that exceeding 80 characters is a lesser evil than not following the indentation level.

Blackladder replied:
should be fixed now

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1895,6 +1925,41 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
+	if (atomic_test_and_set_bit(conn->flags,
+				    BT_CONN_PENDING_VERSION_INFO)) {
+		return -EBUSY;
+	}
```
Empty line after `{`

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -249,6 +249,13 @@ config BT_AUTO_PHY_UPDATE
+	  Enable this to get access to the remote version through
+	  bt_conn_get_info(). The fields in question in the bt_conn_info struct
+
+
```
Remove one of these empty lines (there should never be a need to have more than one consecutive one)

jhedberg reviewed (changes requested):
Almost there! Just a few more minor cleanups and then we can merge. Thanks for your patience with the reviews!

Btw, what's the status with then controller side? Was that already fixed, or we still need to wait for it before being able to merge this PR? @cvinayak ?

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1643,6 +1646,12 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 			if (conn->type == BT_CONN_TYPE_LE) {
 				k_delayed_work_cancel(&conn->update_work);
+#if defined(CONFIG_BT_REMOTE_VERSION)
+				if (atomic_test_and_clear_bit(conn->flags,
```
Actually, why are you doing this in a `if( conn->type == BT_CONN_TYPE_LE)` branch? I don't think we limited the version request to LE-only connections. You'll also end up with a shorter line if you move it out from that branch.

Update: I now see that I overlooked this as well - you seem to be doing it only for LE even though it's not an LE-specific procedure. See my other comments.

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1895,6 +1925,42 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
 }
 
+#if defined(CONFIG_BT_REMOTE_VERSION)
+static int hci_le_read_remote_version(struct bt_conn *conn)
```
This is not an LE-specific operation, so drop the "le" from the function name

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1884,6 +1896,24 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
+			info->manufacturer = conn->rv.manufacturer;
+			info->subversion = conn->rv.subversion;
+		}
+#endif /* CONFIG_BT_REMOTE_VERSION */
```
Sorry for not realising this earlier, but this whole block should apply both for BR/EDR connections as well as LE connections, so pleas move it outside (before) the switch statement.

joerchan commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1895,6 +1925,42 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
+				sizeof(*cp));
+	if (!buf) {
+		atomic_clear_bit(conn->flags, BT_CONN_PENDING_VERSION_INFO);
+		k_sem_give(&conn->rv.sem);
```
Why k_sem_give here? 
I don't see why it is needed here. Maybe I'm missing something @jhedberg?
Same for the one below.

Blackladder replied:
If there is an error or no buffer, we won't actually get to the process of retrieving the remote version and as such it won't actually be pending and we clear the flag.

joerchan replied:
I understand the flag, I'm asking about the k_sem_give. We haven't acquired it.

Blackladder replied:
Ah, then to answer you question "why are they there?" I made a mistake that will be corrected now.

joerchan reviewed (changes requested):
Looks good. I only have one question regarding the use of k_sem_give in some places.

Hopefully we can get this merged after that :)

jhedberg commented:
Did you try to build this?
```
subsys/bluetooth/host/conn.c:1891:14: warning: implicit declaration of function 'hci_read_remote_version' [-Wimplicit-function-declaration]
    int err = hci_read_remote_version(conn);
              ^~~~~~~~~~~~~~~~~~~~~~~
subsys/bluetooth/host/conn.c: At top level:
subsys/bluetooth/host/conn.c:1930:12: error: static declaration of 'hci_read_remote_version' follows non-static declaration
 static int hci_read_remote_version(struct bt_conn *conn)
            ^~~~~~~~~~~~~~~~~~~~~~~
subsys/bluetooth/host/conn.c:1891:14: note: previous implicit declaration of 'hci_read_remote_version' was here
    int err = hci_read_remote_version(conn);
              ^~~~~~~~~~~~~~~~~~~~~~~
subsys/bluetooth/host/conn.c: In function 'hci_read_remote_version':
subsys/bluetooth/host/conn.c:1951:8: error: 'null' undeclared (first use in this function)
        null);
        ^~~~
subsys/bluetooth/host/conn.c:1951:8: note: each undeclared identifier is reported only once for each function it appears in
subsys/bluetooth/host/conn.c:1957:28: error: 'HCI_CMD_TIMEOUT' undeclared (first use in this function); did you mean 'LE_CONN_TIMEOUT'?
  k_sem_take(&conn->rv.sem, HCI_CMD_TIMEOUT);
                            ^~~~~~~~~~~~~~~
                            LE_CONN_TIMEOUT
At top level:
subsys/bluetooth/host/conn.c:1930:12: warning: 'hci_read_remote_version' defined but not used [-Wunused-function]
 static int hci_read_remote_version(struct bt_conn *conn)
            ^~~~~~~~~~~~~~~~~~~~~~~
[172/179] Building C object zephyr/kernel/CMakeFiles/kernel.dir/poll.c.obj
```

Please also enable the new feature for `tests/bluetooth/shell/prj.conf` so that we have CI exercise the new code and don't risk merging something that doesn't build.

To summarise the needed fixes for the above errors:
 - You need to move HCI_CMD_TIMEOUT to hci_core.h
 - You need to use NULL instead of null
 - You need to move hci_read_remote_version above bt_conn_get_info

jhedberg commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1864,12 +1873,33 @@ const bt_addr_le_t *bt_conn_get_dst(const struct bt_conn *conn)
 {
+	if (conn->state != BT_CONN_CONNECTED) {
+		return -ENOTCONN;
+	}
```
Minor: add an empty line after `}`

Blackladder replied:
Why doen't checkpatch detect this?

jhedberg replied:
Because it's not that strict. This is mostly about keeping consistency with the rest of the code in the Bluetooth subsystem.

jhedberg commented on samples/bluetooth/peripheral/prj.conf:
```diff
@@ -22,4 +22,4 @@ CONFIG_FLASH_PAGE_LAYOUT=y
 CONFIG_FCB=y
 CONFIG_SETTINGS=y
-CONFIG_SETTINGS_FCB=y
+CONFIG_SETTINGS_FCB=y
```
Where did this change come from?

joerchan replied:
Looks like an accident from the force-push 1 minute ago.

Blackladder replied:
I add CONFIG_BT_REMOTE_VERSION to the sample and removed slightly too much, should be fixed now.

jhedberg commented:
Once this PR is merged, or even as part of it (as a separate patch), could you also extend the `cmd_info()` function in `subsys/bluetooth/shell/bt.c` to print the new information? That way we can conveniently test the new feature using `tests/bluetooth/shell`

joerchan commented:
I tried to run the shell with this.

I ran into this assert: `assert: '!bt_hci_evt_is_prio(hdr->evt)' failed`

jhedberg commented:
> I tried to run the shell with this.
> I ran into this assert: assert: '!bt_hci_evt_is_prio(hdr->evt)' failed

Yep, that's the known controller bug I've mentioned a few times. I brought it up a week or two ago in slack, and @cvinayak was in the loop at least then

Blackladder commented:
> Btw, what's the status with then controller side? Was that already fixed, or we still need to wait for it before being able to merge this PR? @cvinayak ?
seems like we still need to wait for @cvinayak

joerchan commented:
Right. Should have remembered.

Blackladder commented:
Regarding the changes to `cmd_info()`: do I add something like
```#if defined
        shell_print("Version: %s, manufacturer 0x%4x, subversion: 0x%4x",get_version_str(info.version), info.manufacturer, info.subversion);
```
the manufacturer number list is over 2000, so we might not want to turn that into a string

joerchan commented:
> Regarding the changes to `cmd_info()`: do I add something like
> 
> ```
>         shell_print("Version: %s, manufacturer 0x%4x, subversion: 0x%4x",get_version_str(info.version), info.manufacturer, info.subversion);
> ```
> 
> the manufacturer number list is over 2000, so we might not want to turn that into a string

Yeah, printing the hex number is good enough. You can use 0x%04x so that it shows up as "0x0abc".

Blackladder commented:
Is there an existing function for LL number to string or will I have to make it? If I have to make it, where should I put it?

joerchan commented:
> Is there an existing function for LL number to string or will I have to make it? If I have to make it, where should I put it?

Just print the number for now. We can that in a different PR. Let's get this one done.

joerchan commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1864,12 +1873,68 @@ const bt_addr_le_t *bt_conn_get_dst(const struct bt_conn *conn)
 	info->id = conn->id;
 
+#if defined(CONFIG_BT_REMOTE_VERSION)
+		if (atomic_test_bit(conn->flags, BT_CONN_HAVE_VERSION_INFO)) {
```
This is at the wrong indentation

Commit:
Bluetooth: shell: Add printing of remote version information

Add printing of the remote version information whenever the new
CONFIG_BT_REMOTE_VERSION option is enabled.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: shell: Fix floating point usage

Avoid floating point usage when possible. E.g. on qemu_x86 this would
otherwise result in something like the following:

<err> os: Floating point unit not enabled
<err> os: eax: 0x00000024, ebx: 0x00000000, ecx: 0x00248fc8,
          edx: 0x00146120
<err> os: esi: 0x0012b30c, edi: 0x0024c628, ebp: 0x0024c638,
          esp: 0x0024c5b8
<err> os: eflags: 0x00000216 cs: 0x0008 cr3: 0x001465a0
<err> os: call trace:
<err> os: eip: 0x00121c8d
<err> os:      0x0010679c (0x127750)
<err> os:      0x0010693b (0x100f33)
<err> os:      0x001061bc (0x12b800)
<err> os:      0x001074fb (0x0)
<err> os:      0x00101151 (0x12b800)
<err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception
<err> os: Current thread: 0x00146120 (shell_uart)
<err> os: Halting system

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
@Blackladder I've pushed some updates of my own (both some fix-up of your patch as well as two new patches from me). If you do further updates please update your local tree from the PR one first. Most notable changes:

 - I enabled the feature for `tests/bluetooth/shell/prj.conf` (I had asked this earlier but seems you missed that comment?)
 - I had to change `bt_conn_get_info()` to *not* fail even if we're not connected. The reason is that we have code that's calling this API from disconnected callbacks (mainly to log the address of the connection)
- I added support for the shell to log the info. I couldn't test with the native controller since it asserts (still waiting for a fix from @cvinayak for this), and when I tested with qemu_x86 there was a floating point fault (hence one of my two patches)

joerchan commented:
@jhedberg Your commit message is too long. gitlint is complaining.

jhedberg commented:
> @jhedberg Your commit message is too long. gitlint is complaining.

That's verbatim console output (actually, I even shortened the lines by removing timestamps, but further shortening would loose essntial information). This is in particular case where gitlint must be ignored.

joerchan commented:
@jhedberg @Blackladder I tested with the PTS dongle. Seems to work.

jhedberg commented:
@joerchan @Blackladder @cvinayak  I'll keep my "request changes" until we have a fix for the native controller in this PR. The reason is that the shell isn't usable now with it due to the shell implicitly calling `bt_conn_get_info()` from the `connected` callback, and thereby triggering the assert.

joerchan commented:
@jhedberg Couldn't we disable the feature in the shell and get it merged?

jhedberg commented:
> @jhedberg Couldn't we disable the feature in the shell and get it merged?

Then we'd need to enable it in some other sample, since it's essential that CI exercises the new code and complains if something is broken with it.

Honestly, I'd prefer to wait for the native controller fix - I fear the bug might either get forgotten or given a lower priority otherwise. What's preventing it from getting fixed e.g. today? I doubt it should require more than a couple of lines changed? (that said, I'm not enough of a controller expert to say for sure)

carlescufi commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -249,6 +249,12 @@ config BT_AUTO_PHY_UPDATE
 	  want to rely on remote device to initiate the procedure at its
 	  discretion.
 
+config BT_REMOTE_VERSION
```
@jhedberg we should move this to the common Kconfig settings so that I can refer to it in https://github.com/zephyrproject-rtos/zephyr/pull/19657

jhedberg replied:
@carlescufi What exactly do you want to do on the controller side? Don't we want this always enabled for split (controller only) builds? AFAIK it's a mandatory command for the controller whenever connections are supported.

carlescufi replied:
I meant enabling the processing in the priority thread. But in fact what I'll do is to conditionally compile it based on whether it's a combined build or not, so let's not move it. Sorry for the noise.

carlescufi commented:
@jhedberg the fix for the controller is here: https://github.com/zephyrproject-rtos/zephyr/pull/19657

carlescufi reviewed (changes requested):
Please wait until https://github.com/zephyrproject-rtos/zephyr/pull/19657 is merged and adapt to that particular PR.

joerchan committed:
Bluetooth: host: Refactor host auto initiated LL procedures

Refactor the handling of the host auto initiated LL procedures.
This makes it easier to add new auto initiated procedures as well as
reduced the maintenance by reducing code duplication.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: Change remote version event from prio to normal event type

This commit reverts the change that moved the remote version event from
a priority event to a normal event. This is done because the strategy
for using this event has been changed and will be used with a callback
instead of a semaphore that could be locked from the RX thread.

This commit retains the infrastructure that was added in the controller
so that moving events to priority processing is still possible.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -967,6 +965,33 @@ static int hci_le_read_remote_features(struct bt_conn *conn)
+	}
+
+	/* Remote version cannot change. */
+	if (atomic_test_and_set_bit(conn->flags, BT_CONN_AUTO_VERSION_INFO)) {
```
Remove and set

carlescufi commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -967,6 +965,33 @@ static int hci_le_read_remote_features(struct bt_conn *conn)
 	return 0;
 }
 
+int hci_read_remote_version(struct bt_conn *conn)
```
Should this be enclosed in `#ifdef CONFIG_BT_REMOTE_VERSION` ?

joerchan replied:
It should be static atleast.

I think using IS_ENABLED(CONFIG_BT_REMOTE) { hci_read_remote_version() } is enough to optimize it out. So it should not be included this way.
Checking with `nm` there is no hci_read_remote_version in the zephyr.elf file.

carlescufi commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3231,6 +3269,37 @@ static void hci_encrypt_key_refresh_complete(struct net_buf *buf)
+		conn->rv.subversion = sys_le16_to_cpu(evt->subversion);
+	}
+
+	atomic_set_bit(conn->flags, BT_CONN_AUTO_VERSION_INFO);
```
Are you using `BT_CONN_AUTO_VERSION_INFO` before defining it or am I seeing the wrong commit order?

carlescufi reviewed (changes requested):
Looking much better. I think this solution is the right one, instead of bundling this into con info. Also the callback is much better and optimal than the blocking call.
A couple of loose ends.

jhedberg commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -241,6 +284,9 @@ static struct bt_conn_cb conn_callbacks = {
 	.security_changed = security_changed,
 #endif
+#if defined(CONFIG_BT_REMOTE_INFO)
+	.remote_info_available = remote_info_available
```
Add a comma to the end of this line, in case the struct gets extended in the future

joerchan replied:
Good point. 👍

carlescufi reviewed (approved):
Looks great now.
Also, this PR reduces in 105 bytes the ROM footprint of `samples/bluetooth/peripheral` in its default config.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4913,6 +4971,11 @@ static const struct event_handler prio_events[] = {
+	EVENT_HANDLER(BT_HCI_EVT_REMOTE_VERSION_INFO,
+		      bt_hci_evt_read_remote_version_complete,
+		      sizeof(struct bt_hci_evt_remote_version_info)),
+#endif /* CONFIG_BT_REMOTE_VERSION */
```
The handler should either be in `normal_events[]` or in `prio_events[]` but not both. I guess you're "deprioritizing" it along with this PR, so the above chunk should be removed.

joerchan replied:
Thanks. That is an oversight. Will remove.

jhedberg commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -6,6 +6,8 @@
  */
 
+#define HCI_CMD_TIMEOUT		K_SECONDS(10)
+
```
Isn't the move of this define unnecessary now?

joerchan replied:
Yes, I will move it back again.

joerchan reviewed (dismissed):
Revert prio changes not complete,

dbkinder reviewed (approved):
Looks like doc changes are all in the doxygen comments, LGTM.

joerchan commented:
@jhedberg @carlescufi Fixed reverting of the prio changes. I had some trouble with git giving me the wrong commit in blame which ended up silently bringing the changes back. 

So please give this a thorough overview :) in case other cases of amending commits went wrong.

I have looked over now and it looks ok to me. But I don't think it is right that I approve my own changes so I won't approve this.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3814,6 +3814,11 @@ static const struct event_handler normal_events[] = {
+	EVENT_HANDLER(BT_HCI_EVT_REMOTE_VERSION_INFO,
+		      bt_hci_evt_read_remote_version_complete,
+		      sizeof(struct bt_hci_evt_remote_version_info)),
+#endif /* CONFIG_BT_REMOTE_VERSION */
```
It seems that you haven't introduced the  `bt_hci_evt_read_remote_version_complete()` yet in this patch - it only comes in a later one. Btw, a good way to check that each commit builds on its own is something like `git rebase --exec="west build" origin/master`. That will run the command given to the `--exec` switch for every commit that's part of the rebase.

jhedberg replied:
Actually, you haven't introduced the `CONFIG_BT_REMOTE_VERSION` option either here. I know this is a rather pedantic approach but it helps later when doing something like git bisect since every commit that it might land on should be sane and buildable.

joerchan replied:
Thanks, good catch. I don't know exactly how it happened but some of my fixup commits ended up in the wrong commits.
The  `--exec=""` was a good tip. Although in this case it wouldn't have caught it as it still compiled due to the CONFIG_BT_REMOTE_VERSION not having been defined.

Moved the change to the right commit.

dnsglk commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1078,12 +1105,61 @@ static struct bt_conn *find_pending_connect(bt_addr_le_t *peer_addr)
+	    ((conn->role == BT_HCI_ROLE_MASTER) ||
+	     BT_FEAT_LE_SLAVE_FEATURE_XCHG(bt_dev.le.features))) {
+		err = hci_le_read_remote_features(conn);
+		if (!err) {
```
Hi guys,

I wonder whether the condition is correct. Shouldn't it be just `if (err)` ? It looks like it was borrowed from the commits this PR was based on.

I found this issue when tried to update connection parameters from the app level. They are simply not updated, because of the timer-specifics involved in the call of slave_update_conn_param() which is below in this function. And this `slave_update_conn_param()` is not called when conditions like I mentioned return from the `conn_auto_initiate()`..

joerchan replied:
`if (!err)` is the correct expression here. It returns on success off sending the command. If not success it moves to the next one. You can see it goes for all of the HCI procedures in this function. Once the procedure completes this function will be called and the next procedure should be initiated.

joerchan replied:
@dnsglk The intention is to call all of them. But the three first in the list  are asynchronous, and only one can be active at a time.
If `hci_le_read_remote_features` is called, then `conn_auto_initiate` will be called again from `bt_hci_evt_read_remote_version_complete`, and this time it will call the next function in your list. 
`hci_le_set_data_len()` and `slave_update_conn_param()` can be called simultaneously, so they don't have this behavior.

dnsglk replied:
I see now. In my case then, gdb shows that `le_phy_update_complete()` is not called for whatever reason...

dnsglk replied:
@joerchan thanks for spending time explaining it. Anyway, it still looks like something to improve IMO. Because, if remote doesn't respond to that query (which is my guess on why the phy callback is not triggered), it still means that not all of auto initialization has been completed.

joerchan replied:
@dnsglk Can you create an bug-report with the missing `le_phy_update_complete` and tell us how to reproduce this problem? Please be as detailed as possible

dnsglk replied:
@joerchan  I'll post a report if it I have more time to dig deeper into the issue. Thanks for understanding.
